### PR TITLE
Image-Editor: Enable image-editor for Jetpack sites

### PR DESCRIPTION
**Pull request by retrofox**
This PR enables the image-editor for those Jetpack sites which have a version greater or equal than `4.7-alpha`.

### Testing

Try to edit an image for sites with different Jetpack versions. Depending on this, the `Edit Image` button should be shown or hidden. 

**Jetpack version greater or equal to `4-7-alpha`

**Not valid Jetpack site**

### Testing task

- [x] Do not show for private sites
- [x] Do not show is the feature flag is not enabled
- [x] Do not show if there is not a valid item yet
- [x] Do not show for non-valid Jetpack version
- [x] Show for valid Jetpack version

--------
_`updated: 2017-02-28` by @retrofox_

### CODE REVIEW

File: client/post-editor/media-modal/detail/detail-item.jsx [hunk 1]
```diff
@@ -113,7 +113,6 @@ class EditorMediaModalDetailItem extends Component {
 		if (
 			site.is_private ||
 			! config.isEnabled( 'post-editor/image-editor' ) ||
-			isJetpack( site ) ||
```
Comment (gwwar) [inline client/post-editor/media-modal/detail/detail-item.jsx]: We'll need to update this to something similar to:
```
isJetpack( site ) && jetpackVersion > minSupportedJetpackImageEditorVersion
```

File: client/post-editor/media-modal/detail/detail-item.jsx [hunk 2]
```diff
@@ -19,9 +19,25 @@ import EditorMediaModalDetailPreviewAudio from './detail-preview-audio';
...
+ * and its version is greater than the minimum version
+ * to be able to edit media files.
+ *
+ * @param  {object} site - current site
+ * @return {boolean} `true` is it's a valid Jetpack site
+ */
+const allowForJetpackSite = site => (
+	site &&
+	isJetpack( site ) &&
+	site.options &&
```
Comment (gwwar) [inline client/post-editor/media-modal/detail/detail-item.jsx]: Maybe some lodash get usage here can reduce the number of undefined checks we need to make.
```jsx
const isJetpack = get( site, 'jetpack', false );
if ( ! isJetpack ) {
    return false;
}
const jetpackVersion = get( site, 'options.jetpack_version', '0.0' );
return versionCompare( jetpackVersion, '4.7-alpha', '>=' );
```

File: client/post-editor/media-modal/detail/detail-item.jsx [hunk 3]
```diff
@@ -19,9 +19,25 @@ import EditorMediaModalDetailPreviewAudio from './detail-preview-audio';
...
 
+/**
+ * This function returns true if the site is a Jetpack site
+ * and its version is greater than the minimum version
+ * to be able to edit media files.
+ *
+ * @param  {object} site - current site
+ * @return {boolean} `true` is it's a valid Jetpack site
+ */
+const allowForJetpackSite = site => (
```
Comment (gwwar) [inline client/post-editor/media-modal/detail/detail-item.jsx]: Maybe we should update this method to just check if the image editor is enabled for all cases?

Right now the logic is being added incorrectly with the other cases, and so the image editor is not visible to normal wpcom sites.

File: client/post-editor/media-modal/detail/detail-item.jsx (hunk 2 above)
Comment (retrofox) [inline client/post-editor/media-modal/detail/detail-item.jsx]: We'll leave to use isJetpack lib util function, which is ok for me.

**Comment by retrofox** [comment]
Thanks @gwwar. The last approach was a shame. I'm sorry. Could you take a look now?
I've created a unique function to check if the image-editor could be enabled for the site; trying to be as clear as possible about the conditions which define the boolean result that it returns.

File: client/post-editor/media-modal/detail/detail-item.jsx [hunk 4]
```diff
@@ -18,10 +19,47 @@ import EditorMediaModalDetailPreviewVideo from './detail-preview-video';
...
+	if ( site.is_private ) {
+		return false;
+	}
+
+	// do not allow for Jetpack site with a non-valid version
+	const isJetpack = get( site, 'jetpack', false );
+	const jetpackVersion = get( site, 'options.jetpack_version', '0.0' );
+	const isValidVersion = versionCompare( jetpackVersion, '4.7-alpha', '>=' );
+
+	if ( isJetpack && ! isValidVersion ) {
```
Comment (artpi) [inline client/post-editor/media-modal/detail/detail-item.jsx]: The above are going to get computed for non-jetpack sites which is bit of a waste.
I would:

```
if (
    get( site, 'jetpack', false ) &&
    versionCompare( get( site, 'options.jetpack_version', '0.0' ) , '4.7-alpha', '>=' )
) {
```

That is not a big clause and we dont have to compute versions for the vast majority of sites that are in calypso

File: client/post-editor/media-modal/detail/detail-item.jsx [hunk 5]
```diff
@@ -18,10 +19,47 @@ import EditorMediaModalDetailPreviewVideo from './detail-preview-video';
...
 
+/**
+ * This function return true if the image editor can be
+ * enabled/shown
+ *
+ * @param  {object} item - media item
+ * @param  {object} site - current site
+ * @return {boolean} `true` is the image-editor can be enabled.
+ */
+const enableImageEdition = ( item, site ) => {
```
Comment (artpi) [inline client/post-editor/media-modal/detail/detail-item.jsx]: I would go with `enableImageEditing`

File: client/post-editor/media-modal/detail/detail-item.jsx (hunk 4 above)
Comment (retrofox) [inline client/post-editor/media-modal/detail/detail-item.jsx]: fair enough

File: client/post-editor/media-modal/detail/detail-item.jsx [hunk 6]
```diff
@@ -18,10 +19,46 @@ import EditorMediaModalDetailPreviewVideo from './detail-preview-video';
...
+		return false;
+	}
+
+	// do not show if the feature flag isn't set
+	if ( ! config.isEnabled( 'post-editor/image-editor' ) ) {
+		return false;
+	}
+
+	// do not allow for private sites
+	if ( site.is_private ) {
```
Comment (gwwar) [inline client/post-editor/media-modal/detail/detail-item.jsx]: Careful, site might be null here. `site && site.is_private` or `get( site, 'is_private' )`

Comment (gwwar) [inline client/post-editor/media-modal/detail/detail-item.jsx]: After bulletproofing this we should be able to 🚢  

I verified behavior on a public wpcom site, private site, Jetpack site, and double checked the version compare behavior.

**Comment by retrofox** [comment]
Thanks Kerry and Artur for the review.

**Pull request merged**